articles/WideStrings.md

```diff
@@ -0,0 +1,213 @@
+
+## Introducing Wide Strings
+ROS 2 messages will have a new [primitive field type](/articles/interface_definition.html) `wstring`.
+This purpose is to allow ROS 2 nodes to comminicate with non-ROS DDS entities using an IDL containing a `wstring` field.
+The encoding of data in this type should be UTF-16 to match DDS-XTYPES 1.2.
+Since both UTF-8 and UTF-16 can encode the same code points, new ROS 2 messages should prefer `string` over `wstring`.
+
+## Encodings are Recommendations
+The choice of UTF-8 or UTF-16 for `string` and `wstring` are recommendations.
+A `string` or `wstring` field is allowed to be populated and published with an unknown encoding.
+If subscriber receives a string containing an unknown ncoding then it is up to the subscriber how to handle it.
+It could decode it using an encoding that has been agreed to out of band.
+Other subscribers like `ros2 topic echo` may echo the bytes in hexadecimal.
+
+In practice using different encodings may not be possible.
+The IDL specification forbids `string` from containing `NULL` values.
+For compatibility a ROS message `string` field must not contain zero bytes, and a `wstring` field must not contain zero words.
+An encoding like UTF-32 cannot be used since it may have zero bytes or words in a code point.
+If UTF-32 encoded strings need to be transmitted then one should either convert to UTF-8 and use a `string` or use a `uint32` array.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> Can ROS 1 publish a string with NULL bytes?
+</div>
+
+## Unicode Strings Across ROS 1 Bridge
+Since ROS 1 and 2 both allow `string` to be UTF-8, the ROS 1 bridge will pass values unmodified between them.
+If a message sent from ROS 2 to ROS 1 fails to serialize because a string contains values that are not legal UTF-8 then it won't be published by the bridge.
+
+If a ROS 2 message has a field of type `wstring` then the bridge will attempt to convert it from UTF-16 to UTF-8.
+The resulting UTF-8 encoded string will be published as a `string` type.
+If the conversion fails then the bridge will not publish the message.
+
+## Size of a Wide String
+
+Both UTF-8 and UTF-16 are variable width encodings.
+To minimize the amount of memory used, the `string` and `wstring` types are to be stored in client libraries according to the smallest possible code point.
+This means `string` must be specified as a sequence of bytes, and `wstring` is to be specified as a sequence of words.
+
+Some DDS implementations use 32bit types to store wide strings values.
+This may be due to DDS-XTYPES 1.1 spec saying `wchar` is a 32bit value.
+ROS 2 will instead aim to be compatible with DDS-XTYPES 1.2 and use 16bit storage for wide characters.
+Generated code for ROS 2 messages will automatically handle the conversion when a message is serialized or deserialized.
+
+### Bounded wide strings
+
+Embedded systems accepting strings may use message definitions that restrict the maximum size of a string.
```

Comment:
I would remove the "embedded systems" part since this happens in the message definition and affect all platforms.

Reply:
e4cd96a90926ba81071364d9c1576683adebf565